README.md
# koa-session-mongo

[![Build Status](https://secure.travis-ci.org/hiddentao/koa-session-mongo.png)](http://travis-ci.org/hiddentao/koa-session-mongo) [![NPM module](https://badge.fury.io/js/koa-session-mongo.png)](https://npmjs.org/package/koa-session-mongo)

MongoDB storage layer for [Koa session middleware](https://github.com/hiddentao/koa-session-store). Based on [connect-mongo](https://github.com/kcbanner/connect-mongo).

## Installation

```bash
npm install koa-session-mongo
```

## Usage

```js
var session = require('koa-session-store');
var mongoStore = require('koa-session-mongo');
var koa = require('koa');

var app = koa();

app.keys = ['some secret key'];  // needed for cookie-signing

app.use(session({
  store: mongoStore.create({
    db: 'database_name'
  })
}));

app.use(function *(){
    var n = this.session.views || 0;
    this.session.views = ++n;
    this.body = n + ' views';
});

app.listen(3000);
console.log('listening on port 3000');
```

If you wish to specify host, port, etc:

```js
mongoStore.create({
  host: 'mongo.hostname.com',
  port: 48473,
  db: 'database_name',
  username: 'admin',
  password: 'password',
})
```

Or you can pass in connection parameters as a URL string:

```js
mongoStore.create({
  url: 'mongodb://user:pass@host:port/database_name/collection_name'
})
```

Or you can use an existing [node-mongo-native](https://github.com/mongodb/node-mongodb-native) db object:

```js
var mongo = require('mongodb');

var db = new mongo.Db("database_name", new mongo.Server('host', port, {}), { w: 1 });

mongoStore.create({
  db: dbConn,
  collection: 'sessions',
  username: 'admin',
  password: 'password'
})
```

Or you can use an existing [Mongoose](https://github.com/LearnBoost/mongoose) connection:

```js
var mongoose = require('mongoose');

mongoose.connect(...);

var store = mongoStore.create({
  mongoose: mongoose.connection
})
```


## Options

The following configuration options are available for the `create()` call:

  * **db** `String` or `Object` - db name or instantiated [node-mongo-native](https://github.com/mongodb/node-mongodb-native) db object.
  * **collection** `String` - collection name. Default is _sessions_.
  * **host** `String` server hostname. Default is _127.0.0.1_.
  * **port** `Number` - server port. Default is _27017_.
  * **username** `String` - username. Default is _null_.
  * **password** `String` - password. Default is _null_.
  * **auto_reconnect** `Boolean` - gets passed to the [node-mongo-native](https://github.com/mongodb/node-mongodb-native) constructor as the same option. Default is _false_.
  * **ssl** `Boolean` - use ssl to connect to the server. Default is _false_.
  * **expirationTime** `Number` - time-to-live (TTL) in seconds for any given session data - MongoDB will auto-delete data which hasn't been updated for this amount of time. Default is 2 weeks.
  * **url** `String` - connection URL of the form** `mongodb://user:pass@host:port/database/collection`. If provided then this will take precedence over other options except** `mongoose`.
  * **mongoose** `Object` - a [Mongoose](https://github.com/LearnBoost/mongoose) connection, use** `mongoose.connection` to get the connection out of an existing Mongoose object.  If provided then this will take precedence over other options.

## Cleaning up

You can close all connections made through the storage layer using:

```js
var Promise = require('bluebird');
var mongoStore = require('koa-session-mongo');

...

Promise.spawn(function() {
  mongoStore.closeConnections();
});
```

## License

Copyright (c) 2013 [Ramesh Nair](http://hiddentao.com/)

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.


CHANGELOG.md
# Changelog for [koa-session-mongo](https://github.com/hiddentao/koa-session-mongo)

## 22 Dec 2013 (0.1.1)
* Rewrote code so that actual db connection is only made when session gets used, making initialisation synchronous.

## 22 Dec 2013 (0.1.0)
* Initial version


test.js
var _ = require('lodash');

require("mocha-as-promised")();

var chai = require('chai');
chai.use(require("chai-as-promised"));

var expect = chai.expect,
  assert = chai.assert,
  should = chai.should();

var Promise = require('bluebird');

var mongoSession = require('./');
var createSession = Promise.coroutine(mongoSession.create),
  closeConnections = Promise.coroutine(mongoSession.closeConnections);

var mongo = require('mongodb');
var mongoose = require('mongoose');


describe('Mongo session layer tests', function() {
  var testDb, testCollection;

//  var options = {
//    host: '127.0.0.1',
//    port: 27017,
//    db: 'koa-session-mongo-test'
//  };
//

  before(function(done) {
    console.log('Assuming that Mongod is running on 127.0.0.1 at port 27017');
    var dbConn = new mongo.Db("koa-session-mongo-test", new mongo.Server('127.0.0.1', 27017, {}), { w: 1 });

    Promise.promisify(dbConn.open, dbConn)()
      .then(function(openedDb) {
        testDb = openedDb;
        return Promise.promisify(testDb.collection, testDb)('sessions');
      })
      .then(function(collection) {
        testCollection = {
          findOne : Promise.promisify(collection.findOne, collection),
          update : Promise.promisify(collection.update, collection),
          remove: Promise.promisify(collection.remove, collection)
        };
      })
      .nodeify(done)
    ;
  });

  after(function(done) {
    testDb.close(true, done);
  });


  afterEach(function(done) {
    closeConnections().nodeify(done);
  });


  describe('when no options given', function() {
    it('throws', function() {
      return createSession().should.be.rejectedWith('Missing option: db');
    });
  });

  describe('when native db object given', function() {
    var testMongoNativeDb, options;

    beforeEach(function() {
      testMongoNativeDb = new mongo.Db("koa-session-mongo-test", new mongo.Server('127.0.0.1', 27017, {}), { w: 1 });
      options = {
        db: testMongoNativeDb
      };
    });

    afterEach(function(done) {
      Promise.promisify(testMongoNativeDb.close, testMongoNativeDb)(true).nodeify(done);
    });

    describe('normally', function() {

      it('connects successfully', function() {
        return createSession(options).should.be.fulfilled;
      });

      it('returns the store', function(done) {
        createSession(options)
          .then(function(store) {
            store.save.should.be.a('function');
            store.load.should.be.a('function');
            store.remove.should.be.a('function');
          })
          .nodeify(done)
        ;
      });

      describe('once the store is ready', function() {
        var store, sid;

        beforeEach(function(done) {
          sid = mongoose.Types.ObjectId();

          createSession(options)
            .then(function(_store) {
              store = {
                save: Promise.coroutine(_store.save).bind(_store),
                load: Promise.coroutine(_store.load).bind(_store),
                remove: Promise.coroutine(_store.remove).bind(_store)
              };
            })
            .then(function() {
              return testCollection.remove();
            })
            .nodeify(done)
          ;
        });

        it('can save new items', function() {
          return store.save(sid, 'data1')
            .then(function() {
              return testCollection.findOne({_id: sid});
            })
            .then(function(foundItem) {
              foundItem.blob.should.eql('data1');
            });
        });

        it('can overwrite items', function() {
          return testCollection.update({ _id: sid }, {$set: {blob: 'hell'}}, {upsert:true, safe:true})
            .then(function() {
              return store.save(sid, 'data1')
            })
            .then(function() {
              return testCollection.findOne({_id: sid});
            })
            .then(function(foundItem) {
              foundItem.blob.should.eql('data1');
            });
        });

        it('can load items', function() {
          return testCollection.update({ _id: sid }, {$set: {blob: 'hell'}}, {upsert:true, safe:true})
            .then(function() {
              return store.load(sid);
            })
            .then(function(data) {
              data.should.eql('hell');
            });
        });

        it('can remove items', function() {
          return testCollection.update({ _id: sid }, {$set: {blob: 'hell'}}, {upsert:true, safe:true})
            .then(function() {
              return store.remove(sid);
            })
            .then(function() {
              return testCollection.findOne({_id: sid});
            })
            .then(function(data) {
              expect(data).to.eql(null);
            });
        });

      });

    });

    describe('when incorrect auth given', function() {
      it('fails to connect', function() {
        return createSession(_.extend({}, options, {
          username: 'admin',
          password: 'password'
        })).should.be.rejectedWith('Error authenticating with admin: auth fails');
      });
    });

  });

  describe('when url given', function() {
    var options;

    beforeEach(function() {
      options = {
        url: 'mongodb://127.0.0.1:27017/koa-session-mongo-test/sessions'
      };
    });

    describe('normally', function() {

      it('connects successfully', function() {
        return createSession(options).should.be.fulfilled;
      });

      it('returns the store', function(done) {
        createSession(options)
          .then(function(store) {
            store.save.should.be.a('function');
            store.load.should.be.a('function');
            store.remove.should.be.a('function');
          })
          .nodeify(done)
        ;
      });

      describe('once the store is ready', function() {
        var store, sid;

        beforeEach(function(done) {
          sid = mongoose.Types.ObjectId();

          createSession(options)
            .then(function(_store) {
              store = {
                save: Promise.coroutine(_store.save).bind(_store),
                load: Promise.coroutine(_store.load).bind(_store),
                remove: Promise.coroutine(_store.remove).bind(_store)
              };
            })
            .then(function() {
              return testCollection.remove();
            })
            .nodeify(done)
          ;
        });

        it('can save new items', function() {
          return store.save(sid, 'data1')
            .then(function() {
              return testCollection.findOne({_id: sid});
            })
            .then(function(foundItem) {
              foundItem.blob.should.eql('data1');
            });
        });

        it('can overwrite items', function() {
          return testCollection.update({ _id: sid }, {$set: {blob: 'hell'}}, {upsert:true, safe:true})
            .then(function() {
              return store.save(sid, 'data1')
            })
            .then(function() {
              return testCollection.findOne({_id: sid});
            })
            .then(function(foundItem) {
              foundItem.blob.should.eql('data1');
            });
        });

        it('can load items', function() {
          return testCollection.update({ _id: sid }, {$set: {blob: 'hell'}}, {upsert:true, safe:true})
            .then(function() {
              return store.load(sid);
            })
            .then(function(data) {
              data.should.eql('hell');
            });
        });

        it('can remove items', function() {
          return testCollection.update({ _id: sid }, {$set: {blob: 'hell'}}, {upsert:true, safe:true})
            .then(function() {
              return store.remove(sid);
            })
            .then(function() {
              return testCollection.findOne({_id: sid});
            })
            .then(function(data) {
              expect(data).to.eql(null);
            });
        });

      });

    });

    describe('when incorrect auth given', function() {
      it('fails to connect', function() {
        return createSession({
          url: 'mongodb://admin:password@127.0.0.1:27017/koa-session-mongo-test/sessions'
        }).should.be.rejectedWith('Error authenticating with admin: auth fails');
      });
    });

  });

  describe('when Mongoose connection given', function() {
    var options;

    beforeEach(function() {
      var mongooseDb = mongoose.connect('mongodb://127.0.0.1:27017/koa-session-mongo-test');
      options = {
        mongoose: mongooseDb.connection
      };
    });

    afterEach(function(done) {
      Promise.promisify(mongoose.disconnect, mongoose)().nodeify(done);
    });

    describe('normally', function() {

      it('connects successfully', function() {
        return createSession(options).should.be.fulfilled;
      });

      it('returns the store', function(done) {
        createSession(options)
          .then(function(store) {
            store.save.should.be.a('function');
            store.load.should.be.a('function');
            store.remove.should.be.a('function');
          })
          .nodeify(done)
        ;
      });

      describe('once the store is ready', function() {
        var store, sid;

        beforeEach(function(done) {
          sid = mongoose.Types.ObjectId();

          createSession(options)
            .then(function(_store) {
              store = {
                save: Promise.coroutine(_store.save).bind(_store),
                load: Promise.coroutine(_store.load).bind(_store),
                remove: Promise.coroutine(_store.remove).bind(_store)
              };
            })
            .then(function() {
              return testCollection.remove();
            })
            .nodeify(done)
          ;
        });

        it('can save new items', function() {
          return store.save(sid, 'data1')
            .then(function() {
              return testCollection.findOne({_id: sid});
            })
            .then(function(foundItem) {
              foundItem.blob.should.eql('data1');
            });
        });

        it('can overwrite items', function() {
          return testCollection.update({ _id: sid }, {$set: {blob: 'hell'}}, {upsert:true, safe:true})
            .then(function() {
              return store.save(sid, 'data1')
            })
            .then(function() {
              return testCollection.findOne({_id: sid});
            })
            .then(function(foundItem) {
              foundItem.blob.should.eql('data1');
            });
        });

        it('can load items', function() {
          return testCollection.update({ _id: sid }, {$set: {blob: 'hell'}}, {upsert:true, safe:true})
            .then(function() {
              return store.load(sid);
            })
            .then(function(data) {
              data.should.eql('hell');
            });
        });

        it('can remove items', function() {
          return testCollection.update({ _id: sid }, {$set: {blob: 'hell'}}, {upsert:true, safe:true})
            .then(function() {
              return store.remove(sid);
            })
            .then(function() {
              return testCollection.findOne({_id: sid});
            })
            .then(function(data) {
              expect(data).to.eql(null);
            });
        });

      });

    });

  });

  describe('when normal options given', function() {
    var options;

    beforeEach(function() {
      options = {
        host: '127.0.0.1',
        port: 27017,
        db: 'koa-session-mongo-test'
      };
    });

    describe('normally', function() {

      it('connects successfully', function() {
        return createSession(options).should.be.fulfilled;
      });

      it('returns the store', function(done) {
        createSession(options)
          .then(function(store) {
            store.save.should.be.a('function');
            store.load.should.be.a('function');
            store.remove.should.be.a('function');
          })
          .nodeify(done)
        ;
      });

      describe('once the store is ready', function() {
        var store, sid;

        beforeEach(function(done) {
          sid = mongoose.Types.ObjectId();

          createSession(options)
            .then(function(_store) {
              store = {
                save: Promise.coroutine(_store.save).bind(_store),
                load: Promise.coroutine(_store.load).bind(_store),
                remove: Promise.coroutine(_store.remove).bind(_store)
              };
            })
            .then(function() {
              return testCollection.remove();
            })
            .nodeify(done)
          ;
        });

        it('can save new items', function() {
          return store.save(sid, 'data1')
            .then(function() {
              return testCollection.findOne({_id: sid});
            })
            .then(function(foundItem) {
              foundItem.blob.should.eql('data1');
            });
        });

        it('can overwrite items', function() {
          return testCollection.update({ _id: sid }, {$set: {blob: 'hell'}}, {upsert:true, safe:true})
            .then(function() {
              return store.save(sid, 'data1')
            })
            .then(function() {
              return testCollection.findOne({_id: sid});
            })
            .then(function(foundItem) {
              foundItem.blob.should.eql('data1');
            });
        });

        it('can load items', function() {
          return testCollection.update({ _id: sid }, {$set: {blob: 'hell'}}, {upsert:true, safe:true})
            .then(function() {
              return store.load(sid);
            })
            .then(function(data) {
              data.should.eql('hell');
            });
        });

        it('can remove items', function() {
          return testCollection.update({ _id: sid }, {$set: {blob: 'hell'}}, {upsert:true, safe:true})
            .then(function() {
              return store.remove(sid);
            })
            .then(function() {
              return testCollection.findOne({_id: sid});
            })
            .then(function(data) {
              expect(data).to.eql(null);
            });
        });

      });

    });

    describe('when incorrect auth given', function() {
      it('fails to connect', function() {
        return createSession(_.extend(options, {
          username: 'admin',
          password: 'password'
        })).should.be.rejectedWith('Error authenticating with admin: auth fails');
      });
    });

  });


  describe('when an entry gets overwritten', function() {
    var store, sid;

    beforeEach(function(done) {
      sid = mongoose.Types.ObjectId();

      createSession({
        db: 'koa-session-mongo-test'
      })
        .then(function(_store) {
            store = {
              save: Promise.coroutine(_store.save).bind(_store),
              load: Promise.coroutine(_store.load).bind(_store),
              remove: Promise.coroutine(_store.remove).bind(_store)
            };
        })
        .then(function() {
          return testCollection.remove();
        })
        .nodeify(done)
      ;
    });

    it('has its timestamp updated', function() {
      var oldTs = null;

      return store.save(sid, 'blah')
        .then(function() {
          return testCollection.findOne({_id: sid});
        })
        .then(function(foundItem) {
          oldTs = foundItem.updatedAt;
        })
        .then(function() {
          var resolver = Promise.defer();

          setTimeout(function() {
            store.save(sid, 'blah2')
              .then(function() {
                resolver.resolve();
              })
              .catch(function(err) {
                resolver.reject(err);
              })
            ;
          }, 50);

          return resolver.promise;
        })
        .then(function() {
          return testCollection.findOne({_id: sid});
        })
        .then(function(foundItem) {
          timeSince = foundItem.updatedAt.getTime() - oldTs.getTime();
          expect(0 < timeSince).to.be.true;
        });
    });

  });

  describe('when an entry gets too old', function() {
    var store, sid;

    beforeEach(function(done) {
      sid = mongoose.Types.ObjectId();

      createSession({
        db: 'koa-session-mongo-test',
        expirationTime: 0
      })
        .then(function(_store) {
          store = {
            save: Promise.coroutine(_store.save).bind(_store),
            load: Promise.coroutine(_store.load).bind(_store),
            remove: Promise.coroutine(_store.remove).bind(_store)
          };
        })
        .then(function() {
          return testCollection.remove();
        })
        .nodeify(done)
      ;
    });

    it('gets automatically deleted', function() {
      var NEED_TO_WAIT = 60;

      this.timeout((NEED_TO_WAIT + 1)* 1000);  // it can take upto 60 seconds to

      /* Auto-delete (mongoDB TTL) only works on 2.2+, so let's check the version before testing */
      var adminDb = testDb.admin();
      return Promise.promisify(adminDb.buildInfo, adminDb)()
        .then(function(buildInfo) {
          var versionTokens = buildInfo.version.split('.');
          if (2 > parseInt(versionTokens[0]) || 2 > parseInt(versionTokens[1])) {
            console.warn('Need MongoDB version 2.2 or above to test TTL. Skipping test.');
            return;
          } else {

            return store.save(sid, 'blah')
              .then(function() {
                var resolver = Promise.defer();

                console.log('Need to wait for upto ' + NEED_TO_WAIT + ' seconds... (see http://docs.mongodb.org/manual/tutorial/expire-data/)');
                setTimeout(function() {
                  resolver.resolve();
                }, NEED_TO_WAIT * 1000);

                return resolver.promise;
              })
              .then(function() {
                return store.load(sid);
              })
              .then(function(data) {
                expect(data).to.be.null;
              });

          }
        });
    });

  });

});



tests/longRunning.test.js
/*
Long-running tests not suitable for Continuous Integration.
 */

var _ = require('lodash');

require("mocha-as-promised")();

var chai = require('chai');
chai.use(require("chai-as-promised"));

var expect = chai.expect,
  assert = chai.assert,
  should = chai.should();

var Promise = require('bluebird');

var mongoSession = require('../');
var createSession = Promise.method(mongoSession.create),
  closeConnections = Promise.coroutine(mongoSession.closeConnections);

var mongo = require('mongodb');
var mongoose = require('mongoose');


describe('Mongo session layer long-running tests', function() {
  var testDb, testCollection;

  before(function(done) {
    console.log('Assuming that Mongod is running on 127.0.0.1 at port 27017');
    var dbConn = new mongo.Db("koa-session-mongo-test", new mongo.Server('127.0.0.1', 27017, {}), { w: 1 });

    Promise.promisify(dbConn.open, dbConn)()
      .then(function(openedDb) {
        testDb = openedDb;
        return Promise.promisify(testDb.collection, testDb)('sessions');
      })
      .then(function(collection) {
        testCollection = {
          findOne : Promise.promisify(collection.findOne, collection),
          update : Promise.promisify(collection.update, collection),
          remove: Promise.promisify(collection.remove, collection)
        };
      })
      .nodeify(done)
    ;
  });

  after(function(done) {
    testDb.close(true, done);
  });


  afterEach(function(done) {
    closeConnections().nodeify(done);
  });

  describe('when an entry gets too old', function() {
    var store, sid;

    beforeEach(function(done) {
      sid = mongoose.Types.ObjectId();

      createSession({
        db: 'koa-session-mongo-test',
        expirationTime: 0
      })
        .then(function(_store) {
          store = {
            save: Promise.coroutine(_store.save).bind(_store),
            load: Promise.coroutine(_store.load).bind(_store),
            remove: Promise.coroutine(_store.remove).bind(_store)
          };
        })
        .then(function() {
          return testCollection.remove();
        })
        .nodeify(done)
      ;
    });

    it('gets automatically deleted', function() {
      var NEED_TO_WAIT = 65;

      this.timeout((NEED_TO_WAIT + 1)* 1000);  // it can take upto 60 seconds to

      /* Auto-delete (mongoDB TTL) only works on 2.2+, so let's check the version before testing */
      var adminDb = testDb.admin();
      return Promise.promisify(adminDb.buildInfo, adminDb)()
        .then(function(buildInfo) {
          var versionTokens = buildInfo.version.split('.');
          if (2 > parseInt(versionTokens[0]) || 2 > parseInt(versionTokens[1])) {
            console.warn('Need MongoDB version 2.2 or above to test TTL. Skipping test.');
            return;
          } else {

            return store.save(sid, 'blah')
              .then(function() {
                var resolver = Promise.defer();

                console.log('Need to wait for upto 60 seconds... (see http://docs.mongodb.org/manual/tutorial/expire-data/)');
                setTimeout(function() {
                  resolver.resolve();
                }, NEED_TO_WAIT * 1000);

                return resolver.promise;
              })
              .then(function() {
                return store.load(sid);
              })
              .then(function(data) {
                expect(data).to.be.null;
              });

          }
        });
    });

  });

});



index.js
/*
MongoDB storage layer for koa-session-store.

Based on https://github.com/kcbanner/connect-mongo
 */

var debug = require('debug')('koa-session-mongo');

var mongo = require('mongodb');
var Promise = require('bluebird');
var url = require('url');


/**
 * Default options
 */

var defaultOptions = {
  host: '127.0.0.1',
  port: 27017,
  collection: 'sessions',
  auto_reconnect: false,
  ssl: false,
  expirationTime: 60 * 60 * 24 * 14,    // 2 weeks
  w: 1
};





/**
 * @param db {Object} MongoDB db object.
 * @param options {Object} connection options.
 * @constructor
 */
var MongoStore = function(dbConn, options) {
  this._dbConn = dbConn;
  this._options = options;
};



/**
 *
 * Setup the connection to the database.
 * @private
 */
MongoStore.prototype._connect = Promise.coroutine(function*() {
  // already done?
  if (this._ready) return;

  var dbConn = this._dbConn,
    options = this._options;

  var openedDb = null,
    collection = null;

  try {
    debug('open db connection');
    openedDb = yield Promise.promisify(dbConn.open, dbConn)();
    openedDbConnections.push(openedDb);
  } catch (err) {
    if (!(err instanceof Error)) {
      err = new Error(String(err));
    }
    err.message = 'Error opening db: ' + err.message;
    throw err;
  }

  if (options.username && options.password) {
    try {
      debug('authenticate with %s:%s', options.username, options.password);
      yield Promise.promisify(openedDb.authenticate, openedDb)(options.username, options.password);
    } catch (err) {
      throw new Error('Error authenticating with ' + options.username + ': ' + err.message);
    }
  }

  var collectionName = defaultOptions.collection || options.collection;
  try {
    debug('open collection %s', collectionName);
    collection = yield Promise.promisify(openedDb.collection, openedDb)(collectionName);
  } catch (err) {
    throw new Error('Error opening collection ' + collectionName + ': ' + err.message);
  }

  try {
    var expirationTime = (undefined !== options.expirationTime ? options.expirationTime : defaultOptions.expirationTime);
    debug('create index on updatedAt with expiry after %d seconds', expirationTime);
    yield Promise.promisify(collection.ensureIndex, collection)({updatedAt: 1}, {expireAfterSeconds: expirationTime});
  } catch (err) {
    throw new Error('Error creating index on ' + collectionName + ': ' + err.message);
  }

  this._findOne = Promise.promisify(collection.findOne, collection);
  this._update = Promise.promisify(collection.update, collection);
  this._remove = Promise.promisify(collection.remove, collection);
  this._ready = true;
});



/**
 * Load data for given id.
 * @param sid {String} session id.
 * @return {String} data.
 */
MongoStore.prototype.load = function*(sid) {
  if (!this._ready) yield this._connect();
  var data = yield this._findOne({_id: sid});
  if (data && data._id) {
    return data.blob;
  } else {
    return null;
  }
};


/**
 * Save data for given id.
 * @param sid {String} session id.
 * @param blob {String} data to save.
 */
MongoStore.prototype.save = function*(sid, blob) {
  if (!this._ready) yield this._connect();
  var data = {
    _id: sid,
    blob: blob,
    updatedAt: new Date()
  };

  yield this._update({_id: sid}, data, { upsert: true, safe: true });
};



/**
 * Remove data for given id.
 * @param sid {String} session id.
 */
MongoStore.prototype.remove = function*(sid) {
  if (!this._ready) yield this._connect();
  yield this._remove({_id: sid});
};


/**
 * List of opened db connections. So that we can cleanup later on.
 * @type {Array}
 */
var openedDbConnections = [];


/**
 * Close all opened db connections.
 */
exports.closeConnections = function*() {
  for (var i in openedDbConnections) {
    var conn = openedDbConnections[i];
    if (conn && conn.close) {
      yield Promise.promisify(conn.close, conn)(true);
    }
  }
};





/**
 * Create a MongoDB store using given options.
 *
 * @param options {Object} options.
 * @param options.db {String|Object} db name or instantiated node-mongo-native db object.
 * @param options.collection {String} collection name. Default is 'sessions'
 * @param options.host {String} server hostname. Default is 127.0.0.1
 * @param options.port {Number} server port. Default is 27017.
 * @param options.username {String} username. Default is null.
 * @param options.password {String} password. Default is null.
 * @param options.auto_reconnect {Boolean} gets passed to the node-mongo-native constructor as the same option. Default is false.
 * @param options.ssl {Boolean} use ssl to connect to the server. Default is false.
 * @param options.expirationTime {Number} time-to-live (TTL) in seconds for any given session data - mongod will auto-delete data
 * which hasn't been updated for this amount of time. Default is 2 weeks.
 *
 * @param options.url {String} connection URL of the form: mongodb://user:pass@host:port/database/collection. If provided then
 * this will take precedence over other options except `mongoose`.
 *
 * @param options.mongoose {Object} a mongoose connection, use mongooseDb.connections[0] to get the connection out of an
 * existing Mongoose object.  If provided then this will take precedence over other options.
 */
exports.create = function(options) {
  options = options || {};
  var dbConn = null;

  // mongoose connection?
  if (options.mongoose) {
    debug('extract params from mongoose connection');
    if (options.mongoose.user && options.mongoose.pass) {
      options.username = options.mongoose.user;
      options.password = options.mongoose.pass;
    }

    dbConn = new mongo.Db(options.mongoose.db.databaseName,
      new mongo.Server(options.mongoose.db.serverConfig.host,
        options.mongoose.db.serverConfig.port,
        options.mongoose.db.serverConfig.options
      ), { w: defaultOptions.w }
    );

  }
  // non-mongoose
  else {
    // url?
    if(options.url) {
      debug('extract params from url %s', options.url);
      var db_url = url.parse(options.url);

      if (db_url.port) {
        options.port = parseInt(db_url.port);
      }

      if (db_url.pathname) {
        var pathname = db_url.pathname.split('/');

        if (pathname.length >= 2 && pathname[1]) {
          options.db = pathname[1];
        }

        if (pathname.length >= 3 && pathname[2]) {
          options.collection = pathname[2];
        }
      }

      if (db_url.hostname) {
        options.host = db_url.hostname;
      }

      if (db_url.auth) {
        var auth = db_url.auth.split(':');

        if (auth.length >= 1) {
          options.username = auth[0];
        }

        if (auth.length >= 2) {
          options.password = auth[1];
        }
      }
    }

    if(!options.db) {
      throw new Error('Missing option: db');
    }

    // options.db already initialised?
    if ('object' === typeof options.db && 'function' === typeof options.db.open) {
      debug('re-using node-mongo-native object');
      dbConn = options.db; // Assume it's an instantiated node-mongo-native Object
    } else {
      options.auto_reconnect = options.auto_reconnect || defaultOptions.auto_reconnect;
      options.ssl = options.ssl || defaultOptions.ssl;

      debug('creating new node-mongo-native object');
      dbConn = new mongo.Db(options.db,
        new mongo.Server(
          options.host || defaultOptions.host,
          options.port || defaultOptions.port,
          options),
        { w: defaultOptions.w }
      );
    } // if options.db not an object
  }

  debug('creating new store');
  return new MongoStore(dbConn, options);
};



